usr/man/man1/mod.1
.TH mod 1
.SH NAME
mod \- Modula-2 compiler
.SH SYNTAX
.B mod
[ options ] name ...
.SH DESCRIPTION
The 
.I Mod
command
compiles one or more Modula-2 programs or implementation modules.
Definition modules are not compiled.
In the absence of options, it will compile all specified modules and link
them together into an executable file called
.I a.out.
.PP
Each program or implementation module must be in a separate file with a name
ending with .mod.
Each definition module must be in a separate file called module.def 
where module is the name of the module.
Object files ending with .o compiled with
.I mod
or some
other compiler may be specified.
.PP
File name arguments ending with .pcd, and .s, are assumed to be \-pcode
and assembly language files respectively,
and are translated and assembled into object files.
.PP
The options available are: 
.TP 12 
.B \-c
Create object files but do not link them together.
.TP 12 
.B \-g
Generate additional symbol table information for the debugger
.IR dbx (1).
.TP 12 
.B \-i
Ignore the errors in some of the modules
and continue compiling the rest of them.
.TP 12 
.BR \-m "flags"
Perform intermodule checking.
If an out-of-date module is encountered, recompile it using the
specified \*(lqflags\*(rq.
The flags are separated by commas or spaces, and must be quoted if spaces
are used.
.TP 12 
.B \-n
Write out what will happen when the same command is entered
without the \*(lq-n\*(rq option.
.TP 12 
.BR \-o name
Create an executable file called \*(lqname\*(rq instead of the default \*(lqa.out\*(rq.
.TP 12 
.B \-pg
Set up object files for profiling by
.IR  gprof (1).
.TP 12 
.B \-r
Retain pcode and assembly language files in the current directory after
compilation.
.TP 12 
.B \-s
Use standard conventions for reserved word case, cardinal data type, and strings
(See Extensions, below).
.TP 12 
.B \-sc
Use standard conventions for cardinal data type (See Extensions, below).
.TP 12 
.B \-sk
Use standard conventions for reserved word case (See Extensions, below).
.TP 12 
.B \-ss
Use standard conventions for string constants (See Extensions, below).
.TP 12 
.B \-u
Convert all identifiers and reserved words to upper case (i.e., ignore
the case of identifiers and reserved words on input).
.TP 12 
.B \-v
Print out messages which state what is occurring during compilation.
.TP 12 
.B \-C
Generate runtime checks for illegal pointers, subrange and index bounds,
and variant record tags.
.TP 12 
.BR \-D "directory"
Use the specified directory for the phases of the compiler and the
location of the standard definition modules and libraries.
.TP 12 
.B \-L
Ignore references to modules
not specified while performing intermodule checking.
(This is useful when checking modules to be placed in a library).
.TP 12 
.B \-M
Perform intermodule checking, but do not recompile if inconsistencies
are found.
.TP 12 
.BR \-N name
Ignore references to the module
name while performing intermodule checking.
(This is useful when the module name is not a Modula-2 module.)
You may use this option as many times as needed.
.TP 12 
.B \-O
Perform code optimizations.
.TP 12 
.B \-P
Stop after generating pcode in a file ending with .pcd. 
.TP 12 
.B \-S
Stop after generating assembly language in a file ending with .s.
.SH "LIBRARY MODULES"
By default, an import of a global module will cause the compiler to look for
the definition module first
in the working directory and then in the standard library directory.
The standard library modules are automatically linked with the program.
.PP
The default may be overridden to specify other directories of definition modules
using the MODPATH environment variable.
MODPATH is set to a sequence of directory names separated by colons.
Those directories will be searched in the order specified to find 
a definition
module.
The corresponding object files or libraries are specified when linking.
The MODPATH environment variable may be set by the user
in .login or in .modpath in the working directory.
If the file .modpath exists in the working directory, the 
.B mod
command will use its first line as the value of the MODPATH variable.
.PP
The following modules are provided by this implementation of Modula-2.
Note that system, memory, io, and bitoperations are builtin modules;
definition modules for them are provided for documentation purposes only.
Only strings and parameters are actually implemented in Modula-2.
.TP 12
system
.br
Built in system module.  Contains types of what word, address, etc., and process
routines.
.TP
memory
.br
Built in storage module.  Sets up pointers properly for runtime checks.
Contains ALLOCATE and DEALLOCATE.
.TP
io
.br
Built in I/O module that provides formatted read and write similar to
.IR scanf (3)
and
.IR printf (3).
.TP
bitoperations
.br
Built in bit manipulation module.
Performs operations such as shift, exclusive or, etc., on integer operands.
.TP
math
.br
Performs mathematical functions.
Interface to the C math library.
.TP
parameters
.br
Accesses command line parameters and environment variables.
.TP
strings
.br
Compares, assigns, and links strings.
.TP
unix
.br
Defines some Unix system calls and C library routines.
.TP
Storage
.br
Standard storage module, for compatibility with standard Modula-2.
Contains ALLOCATE and DEALLOCATE.
.SH "DIFFERENCES AND EXTENSIONS"
This implementation of Modula-2 has compiled and run Wirth's Modula-2 compiler
(as modified by Cambridge University for the VAX) with only minor changes
to make Wirth's compiler more portable.
However, the definition of the language has been relaxed in some areas.
For the most part, these changes are compatible.
.PP
The following is an incomplete list of differences between this compiler and
Wirth's compiler:
.PP
Reserved words and standard identifiers are recognized in upper
and lower case,
Thus, case variations of reserved words may not be used for identifiers.
This feature is disabled by the
.B -sk
option.
.PP
Cardinal and non-negative subranges that do not exceed MAXINT are considered
to be subranges of integer and are compatible with integers.
Subranges that exceed MAXINT are compatible with cardinal and non-negative
subranges.
This feature is disabled by the
.B -sc
option.
.PP
A built in module called
.I io
provides formatted input and output.
The
.I readf
and
.I writef
routines can accept any number of parameters, as long as their types
correspond properly with the format string.
Supported formats include: for integer and cardinal,
.BR d ,
.BR x ,
and
.BR o ;
for real,
.B g
(output only),
.BR f ,
and
.BR e ;
for longreal,
.B G
(output only),
.BR F ,
and
.BR E ;
for char,
.BR c ;
and
for string (array of char),
.B s
and
[] (input only).
.PP
No import of
.I allocate
or
.I deallocate
is required to use new and dispose if the standard memory allocation routines
are desired.
Programs that require checking import 
.I allocate 
and 
.I deallocate
from memory rather than storage.
.PP
The sizes returned by
.I size
and
.I tsize
and expected by
.IR allocate ,
.I deallocate
and
.I newprocess
are in units of bits.
.PP
The
.I system
module includes the type
.IR byte ,
which is analogous to
.IR word ,
as well as appropriate related constants.
There is also a function cputime, which returns the accumulated program
CPU time in milliseconds.
.PP
There is a standard type called
.I longreal
that stores a double precision real value.
A standard function longfloat converts cardinals, integers, or reals 
to longreal.
.PP
Additional standard procedures include:
.TP 12
min(a,b)
Returns the smaller of two cardinal, integer, real, or longreal values.
.TP
max(a,b)
Returns the larger of two cardinal, integer, real, or longreal values.
.TP
assert(condition[,message])
Aborts the program (with the optional message)
if the condition is false.
.TP
number(a)
Returns the number of elements in the specified array.
.TP
first(type)
Returns the smallest legal value of the specified type.
.TP
last(type)
Returns the largest legal value of the specified type.
.PP
Definition modules are not compiled.
.PP
Escape sequences may be placed in strings to specify non-printing characters.
E.g., \en, \et, \er, \ef, \eb, \e\e, \e', and \e" mean
linefeed, tab, carriage return, form feed, backspace, backslash,
single quote, and double quote, respectively.
In addition a \e followed by up to three octal digits specifies the
the character whose ASCII code is the octal value.
A single (double) quote also may be put in a string delimited with
single (double) quotes by specifying two single (double) quotes.
This feature is disabled by the
.B -ss
option.
.PP
The interface to Unix is through a module called
.I unix
rather than the
.I system
module.
The
.I unixcall
procedure is handled for compatibility with the Cambridge compiler,
but is not recommended.
.PP
Additional keywords are recognized in certain contexts.
These keywords are prefixed by @ to avoid conflicting with valid
identifiers.
.TP
Pointer attributes
Attributes may be specified between the keywords
.I pointer
and
.I to
in order to change the default assumptions of Modula-2 pointer with checking.
Recognized attributes are:
.br
.ta 2i
@nocheck	Modula-2 pointer, no checking
.br
@c	C/malloc pointer, no checking
.br
@pascal	Pascal pointer, Pascal checking
.TP
Size and alignment
The size and alignment of data types may be specified preceding any
type specification.
The size and alignment multiples are in bits.
For example,
.br
    type Register = @align 2 @size 4 [-8..7];
.br
defines a type that occupies 4 bits aligned on a multiple of two bits.
.TP
Exports
Exports from a definition module are assumed qualified.  
Unqualified exports are permitted if the @unqualified keyword is used.
Multiple export statements are permitted, but they must occur next to
each other.
.TP
External variables and procedures
A procedure or variable may be accessed by C and Pascal routines using
its unqualified name if the @external attribute
occurs between the keyword procedure and the name of the procedure or
precedes the variable declaration. 
.TP
Uncounted open arrays
Open array parameters appear as two parameters, the address of the array and
the number of element-to-non-Modula-2 programs.
If necessary, the count may be omitted by placing the attribute @nocount
between the keywords
.I array
and
.I of
in the open array declaration.
.SH FILES
.ta 2.6i
file.mod	Program or implementation module
.br
file.def	Definition module
.br
file.pcd	Pcode (-P or -r)
.br
file.s	Assembly code (-S or -r)
.br
/usr/local/lib/mod/mod2.0	Modula-2 compiler front-end
.br
/usr/local/lib/mod/mod2.1	Modula-2 compiler back-end
.br
/usr/local/lib/mod/mod2.2	Intermodule checker
.br
/usr/local/lib/mod/*.def	Standard definition modules
.br
/usr/local/lib/mod/modlib	Default library
.br
/tmp/modNNNNNN.pcd	Temporary Pcode file
.br
/tmp/modNNNNNN.s	Temporary assembly code file
.SH RESTRICTIONS	 
This is an experimental compiler, and thus no warranties are expressed or
implied about its conformance to the definition of the Modula-2 language
or about its proper functioning.
We will endeavor to report and fix bugs, but users should be aware that this
compiler is not a supported product.
.SH DIAGNOSTICS
All error messages suppress subsequent compilation phases.
Error messages ending with a question mark are internal errors, and
probably represent compiler bugs.
When pointer checking is running in a Modula-2 program,
segmentation faults may be generated by the pointer validation test.
These are intentional and should be considered as invalid pointer messages.  
The compiler runs with runtime checks enabled, and may produce core dumps.
Report problems to the author.
.SH "SEE ALSO"
N. Wirth,
.IR "Programming in Modula-2" ,
Springer-Verlag, New York, 1982.
.\"	.SH AUTHOR
.\"	Michael L. Powell
.\"	.br
.\"	Digital Equipment Corporation
.\"	.br
.\"	Western Research Laboratory
.\"	.br
.\"	4410 El Camino Real
.\"	.br
.\"	Los Altos, CA  94022
.\"	.br
.\"	Mail: powell@decwrl.csnet or {decvax,ucbvax}!decwrl!powell
.\"	.PP
.\"	Software and documentation is
.\"	Copyright 1984, Digital Equipment Corporation,
.\"	Maynard, Massachusetts.
.\"	All rights reserved.
.\"	This software is provided under license agreement and must be kept confidential.


usr.bin/modula2/mod/mod.c
/*#@(#)mod.c	1.1	Ultrix	11/28/84*/
/****************************************************************************
 *									    *
 *  Copyright (c) 1984 by						    *
 *  DIGITAL EQUIPMENT CORPORATION, Maynard, Massachusetts.		    *
 *  All rights reserved.						    *
 * 									    *
 *  This software is furnished under a license and may be used and copied   *
 *  only in  accordance with  the  terms  of  such  license  and with the   *
 *  inclusion of the above copyright notice. This software or  any  other   *
 *  copies thereof may not be provided or otherwise made available to any   *
 *  other person.  No title to and ownership of  the  software is  hereby   *
 *  transferred.							    *
 * 									    *
 *  The information in this software is  subject to change without notice   *
 *  and  should  not  be  construed as  a commitment by DIGITAL EQUIPMENT   *
 *  CORPORATION.							    *
 * 									    *
 *  DIGITAL assumes no responsibility for the use  or  reliability of its   *
 *  software on equipment which is not supplied by DIGITAL.		    *
 * 									    *
$Header: mod.c,v 1.4 84/05/19 11:41:54 powell Exp $
 ****************************************************************************/
#include <stdio.h>
#define MODPATH	"MODPATH"	/* environment variable (check "MODPATH=") */
#define MODPATHFILE ".modpath"	/* name of file to source */

#define ARGSIZE 2048
#define FILENAMESIZE	256
#define MAXENVIRONNAMES	1024
#define MAXARGS		1024
int cflag = 0;
int gflag = 0;
int iflag = 0;
int mflag = 0;
int nflag = 0;
int Pflag = 0;
int pflag = 0;
int rflag = 0;
int Sflag = 0;
int tflag = 0;
int vflag = 0;
int rmpcd, rms;
char *ldlistv[MAXARGS] = {0}, **ldlist = ldlistv;
char *modoptv[MAXARGS] = {0}, **modopt = modoptv;
char *mscoptv[MAXARGS] = {0}, **mscopt = mscoptv;
char *ldoptv[MAXARGS] = {0}, **ldopt = ldoptv;
char modpath[ARGSIZE] = "";
char execfileb[FILENAMESIZE] = "a.out", *execfile = execfileb;
char *commandv[MAXARGS], **command;
char filename[FILENAMESIZE];
char mfilename[FILENAMESIZE];
char pfilename[FILENAMESIZE];
char sfilename[FILENAMESIZE];
char ofilename[FILENAMESIZE];
char lfilename[FILENAMESIZE];
char moddir[FILENAMESIZE] = MODLIBDIR;
int usedefaultdir = 0;
char progname[FILENAMESIZE];
extern char **environ;
char *newenv[MAXENVIRONNAMES];
enum filekind {modfile, deffile, pcdfile, asmfile, objfile, libfile};
enum filekind kind;
char *myname;
char *malloc(), *strcpy(), *strcat(), *mktemp();
extern int errno;
extern char *sys_errlist[];

char **addto(l,s) char **l, *s; {
    *l++ = s;
    return(l);
}
char *newname(s) char *s; {
     char *p;
     p = malloc(FILENAMESIZE);
     strcpy(p,s);
     return(p);
}
char *finddot(s) char *s; {
    while(*s) {
	if (*s == '.') return(s);
	s++;
    }
    return(NULL);
}
outfile(saveit,filename,root,ext) int saveit; char *filename, *root, *ext; {
     if (saveit) {
	strcpy(filename,root);
     } else {
	strcpy(filename,"/tmp/mod");
	strcat(filename,"XXXXXX");
	(void) mktemp(filename);
     }
    strcat(filename,ext);
}
initexec(progname) char *progname; {
    command = commandv;
    *command++ = progname;
}
addarg(arg) char *arg; {
    if (arg == 0 || *arg == '\0') {
	fprintf(stderr,"%s: addarg: null string\n",myname);
    } else {
	*command++ = arg;
    }
}
addargv(argv) char *argv[]; {
    while (*argv != 0) {
	*command++ = *argv++;
    }
}
execute(progfile) char *progfile; {
    int pid, status, i;
    *command++ = 0;
    if (vflag || nflag) {
	fprintf(stderr,"%s: ",progfile);
	for (i=0;commandv[i]!=0;i++) {
	    fprintf(stderr,"%s ",commandv[i]);
	}
	fprintf(stderr,"\n");
    }
    status = 0;
    if (!nflag) {
	if ((pid=fork())==0) {
	    execve(progfile,commandv,newenv);
	    fprintf(stderr,"%s: Cannot execute %s (%s)\n",myname,progfile,
		sys_errlist[errno]);
	    exit(1);
	}
	while(wait(&status)!=pid);
	if (vflag) {
	    fprintf(stderr,"%s: %s exited, status=%d\n",myname,progfile,status);
	}
    }
    if (!iflag && status != 0) {
	if ((status & 0xff)== 0) {
	    status = status >> 8;
	}
	exit(status);
    }
    return(status);
}
remove(file) char *file; {
    if (vflag || nflag) {
	fprintf(stderr,"rm %s\n",file);
    }
    if (!nflag) {
	if (unlink(file)!=0) {
	    fprintf(stderr,"%s: Cannot remove %s (%s)\n",myname,file,
		sys_errlist[errno]);
	}
    }
}
nomoreopt(o,argv) char *o, *argv; {
    if (*o != '\0') {
	fprintf(stderr,"%s: %s not understood in option %s\n",myname,o,argv);
	exit(1);
    }
}
main(argc,argv) int argc; char **argv; {
    char *ext, **env, scanfbuff[ARGSIZE];
    int i, status;

    FILE *modpathfile;
    myname = argv[0];
    if (argc <= 1) {
	fprintf(stderr,"usage: %s [options] files.mod\n",myname);
	exit(1);
    }
    env = environ;
    newenv[0] = modpath;
    i = 1;
    while (*env!=NULL) {
	if (strncmp(*env,"MODPATH=",8)==0) {
	    strcpy(modpath,*env);
	} else {
	    newenv[i] = *env;
	    i++;
	}
	env++;
    }
    newenv[i] = NULL;
    modpathfile = fopen(MODPATHFILE,"r");
    if (modpathfile != NULL) {
	if (fscanf(modpathfile," %s ",scanfbuff)!=1) {
	    fprintf(stderr,"Bad .modpath file\n");
	    exit(1);
	}
	strcpy(modpath,"MODPATH=");
	strcat(modpath,scanfbuff);
    }
    if (modpath[0]=='\0') {
	usedefaultdir = 1;
	strcpy(modpath,"MODPATH=.:");
	strcat(modpath,moddir);
    }
    strcat(moddir,"/");
    for(i=1;i<argc;i++) {
	if (argv[i][0]=='-') {
	    switch (argv[i][1]) {
	    case 'c':
		cflag = 1;
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 'P':
		Pflag = 1;
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 'S':
		Sflag = 1;
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 'r':
		rflag = 1;
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 'g':
		gflag = 1;
		modopt = addto(modopt,"-g");
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 'i':
		iflag = 1;
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 'p':
		pflag = 1;
		modopt = addto(modopt,"-pg");
		if (argv[i][2] == 'g') {
		    nomoreopt(&argv[i][3],argv[i]);
		} else {
		    nomoreopt(&argv[i][2],argv[i]);
		}
		break;
	    case 'u':
		modopt = addto(modopt,"-u");
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 'l':
		ldlist = addto(ldlist,argv[i]);
		break;
	    case 'L':
		mscopt = addto(mscopt,"-L");
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 'N':
		mscopt = addto(mscopt,argv[i]);
		break;
	    case 'm':
		mflag = 1;
		mscopt = addto(mscopt,argv[i]);
		break;
	    case 'M':
		mflag = 1;
		break;
	    case 'o':
		nomoreopt(&argv[i][2],argv[i]);
		if (i+1>=argc) {
		    fprintf(stderr,"%s: Missing file for -o\n",myname);
		    exit(1);
		}
		ext = finddot(argv[i+1]);
		if (ext!=NULL) {
		    if (strcmp(ext,".mod")==0) {
			ext = NULL;
		    } else if (strcmp(ext,".def")==0) {
			ext = NULL;
		    } else if (strcmp(ext,".pcd")==0) {
			ext = NULL;
		    } else if (strcmp(ext,".i")==0) {
			ext = NULL;
		    } else if ((ext[1]=='p'||ext[1]=='c'||ext[1]=='o'||
			    ext[1]=='s'|| ext[1]=='f') && ext[2]=='\0') {
			ext = NULL;
		    }
		    if (ext==NULL) {
			fprintf(stderr,"%s: -o would overwrite source file %s\n",
				myname,argv[i+1]);
			exit(1);
		    }
		}
		strcpy(execfile,argv[i+1]);
		i = i+1;
		break;
	    case 'D':
		if (argv[i][2]=='\0') {
		    strcpy(moddir,".");
		} else {
		    strcpy(moddir,&argv[i][2]);
		}
		if (usedefaultdir) {
		    strcpy(modpath,"MODPATH=.:");
		    strcat(modpath,moddir);
		}
		strcat(moddir,"/");
		break;
	    case 'v':
		fprintf(stderr,"%s\n",modpath);
		vflag = 1;
		nomoreopt(&argv[i][2],argv[i]);
		modopt = addto(modopt,"-v");
		mscopt = addto(mscopt,"-v");
		break;
	    case 'n':
		nflag = 1;
		nomoreopt(&argv[i][2],argv[i]);
		break;
	    case 't':
		tflag = 1;
		nomoreopt(&argv[i][2],argv[i]);
		modopt = addto(modopt,"-t");
		break;
	    case 'W':
		mscopt = addto(mscopt,argv[i]);
		break;
	    default:
		modopt = addto(modopt,argv[i]);
		break;
	    }
	} else {
	    processfile(argv[i]);
	}
    }
    if (!Pflag && !Sflag && !tflag) {
	strcpy(lfilename,moddir);
	strcat(lfilename,"modlib");
	if (pflag) {
	    strcat(lfilename,"_p");
	}
	if (mflag) {
	    initexec("mod2.2");
	    addargv(mscoptv);
	    addargv(ldlistv);
	    strcpy(progname,moddir);
	    strcat(progname,"mod2.2");
	    status = execute(progname);
	}
	if (!cflag) {
	    initexec("ld");
	    addargv(ldoptv);
	    addarg("-X");
	    addarg("-o");
	    addarg(execfile);
	    if (pflag) {
		addarg("/usr/lib/gcrt0.o");
	    } else {
		addarg("/lib/crt0.o");
	    }
	    addargv(ldlistv);
	    addarg(lfilename);
	    if (gflag) {
		addarg("-lg");
	    }
	    if (pflag) {
		addarg("-lpc_p");
		addarg("-lm_p");
		addarg("-lc_p");
	    } else {
		addarg("-lpc");
		addarg("-lm");
		addarg("-lc");
	    }
	    status = execute("/bin/ld");
	    if (status != 0) exit(status);
	}
    }
    exit(0);
}
processfile(file) char *file; {
    char *ext, *pcdext, *phase;
    int status;

    if (tflag) {
	pcdext = ".i";
    } else {
	pcdext = ".pcd";
    }
    strcpy(filename,file);
    ext = finddot(filename);
    if (ext == NULL) kind = libfile;
    if (strcmp(ext,".mod")==0) kind = modfile;
    else if (strcmp(ext,".o")==0) kind = objfile;
    else if (strcmp(ext,pcdext)==0) kind = pcdfile;
    else if (strcmp(ext,".s")==0) kind = asmfile;
    else if (strcmp(ext,".def")==0) kind = deffile;
    else kind = libfile;
    
    if (kind != libfile) {
	*ext = '\0';
    }
    switch(kind) {
    case modfile:
	strcpy(mfilename,filename);
	strcat(mfilename,".mod");
	outfile(Pflag||rflag||tflag,pfilename,filename,pcdext);
	outfile(Sflag||rflag,sfilename,filename,".s");
	outfile(1,ofilename,filename,".o");
	break;
    case pcdfile:
	outfile(1,pfilename,filename,pcdext);
	outfile(Sflag||rflag,sfilename,filename,".s");
	outfile(1,ofilename,filename,".o");
	break;
    case asmfile:
	outfile(1,sfilename,filename,".s");
	outfile(1,ofilename,filename,".o");
	break;
    case objfile:
	outfile(1,ofilename,filename,".o");
	break;
    }
    rmpcd = 0;
    rms = 0;
    switch (kind) {
    case modfile:
	if (tflag) {
	    phase = "mod2.0t";
	} else {
	    phase = "mod2.0";
	}
	initexec(phase);
	addargv(modoptv);
	addarg("-o");
	addarg(pfilename);
	addarg(mfilename);
	strcpy(progname,moddir);
	strcat(progname,phase);
	status = execute(progname);
	if (Pflag || status != 0) return;
	rmpcd = !rflag;
    /* fall through to next phase */
    case pcdfile:
	if (tflag) {
	    phase = "mod2.1t";
	} else {
	    phase = "mod2.1";
	}
	initexec(phase);
	addarg(pfilename);
	if (!tflag) {
	    addarg(sfilename);
	}
	strcpy(progname,moddir);
	strcat(progname,phase);
	status = execute(progname);

	if (tflag) return;
	if (rmpcd) {
	    remove(pfilename);
	}
	if (Sflag || status != 0) return;
	rms = !rflag;
    /* fall through to next phase */
    case asmfile:
	initexec("as");
	addarg("-o");
	addarg(ofilename);
	addarg(sfilename);
	status = execute("/bin/as");
	if (rms) {
	    remove(sfilename);
	}
	if (status != 0) return;
    /* fall through to next phase */
    case objfile:
	ldlist = addto(ldlist,newname(ofilename));
	break;
    case libfile:
	ldlist = addto(ldlist,file);
	break;
    }
}
